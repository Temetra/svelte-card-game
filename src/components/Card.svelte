<script lang="ts">
	import type { SvelteComponent } from "svelte/internal";
	import Background from "../cards/CardBackground.svg"
	import { Suit, Rank, getCardComponent } from "~/modules/Cards.ts"

	export let suit: Suit = Suit.Joker;
	export let rank: Rank = Rank.Ace;

	let cardComponent: SvelteComponent = getCardComponent(suit, rank);
</script>

<style type="text/scss">
	section {
		position:relative;

		// Card dimensions with fallback
		width:var(--card-width, 500px);
		height:var(--card-height, 700px);

		:global(svg) {
			position: absolute;
			top:0;
			left:0;
		}
	}
</style>

<section>
	<Background />
	<svelte:component this={cardComponent}/>
</section>