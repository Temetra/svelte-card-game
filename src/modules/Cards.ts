export * from "@/modules/Cards/Card";
export * from "@/modules/Cards/CardState";
export * from "@/modules/Cards/GetCardName";
export * from "@/modules/Cards/Rank";
export * from "@/modules/Cards/Suit";
export * from "@/modules/Cards/Scoring";
