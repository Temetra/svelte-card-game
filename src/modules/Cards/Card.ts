import type { Rank } from "@/modules/Cards/Rank";
import type { Suit } from "@/modules/Cards/Suit";
import type { CardState } from "@/modules/Cards/CardState";

export interface Card {
	rank: Rank,
	suit: Suit
}

export interface StatefulCard extends Card {
	state: CardState
}

export function getCardIndex(card: Card);
export function getCardIndex(rank: Rank, suit: Suit);
export function getCardIndex(x: any, y?: any) {
	if (typeof x == "number" && y != null) {
		// Assume func was given rank and suit
		return y * 100 + x;
	}
	else if (typeof x == "object") {
		// Assume x is a card object
		return x.suit * 100 + x.rank;
	}

	return null;
}