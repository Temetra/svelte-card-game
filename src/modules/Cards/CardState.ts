export enum CardState {
	Hidden = 1 << 0,
	Focused = 1 << 1,
	Flipped = 1 << 2,
	Dealing = 1 << 3,
	Spinning = 1 << 4,
	Discarding = 1 << 5,
}