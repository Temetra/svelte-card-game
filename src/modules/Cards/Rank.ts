export enum Rank {
	Ace = 1,
	Two,
	Three,
	Four,
	Five,
	Six,
	Seven,
	Eight,
	Nine,
	Ten,
	Jack,
	Queen,
	King
}

export const Ranks: Rank[] = Object.keys(Rank).filter(x => Number.isNaN(Number(x))).map(x => Rank[x]);
