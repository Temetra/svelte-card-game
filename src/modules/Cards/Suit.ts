export enum Suit {
	Clubs = 1,
	Diamonds,
	Hearts,
	Spades,
	Joker
}

export const Suits: Suit[] = Object.keys(Suit).filter(x => !Number(x)).map(x => Suit[x]).filter(x => x != Suit.Joker);
