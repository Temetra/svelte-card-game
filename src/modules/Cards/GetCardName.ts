import { Rank } from "@/modules/Cards/Rank";
import { Suit } from "@/modules/Cards/Suit";

export function getCardName(suit: Suit, rank: Rank) {
	if (suit == Suit.Joker) return Suit[suit];
	else return `${Rank[rank]} of ${Suit[suit]}`;
}
