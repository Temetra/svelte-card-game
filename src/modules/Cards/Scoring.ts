import { Card, getCardIndex } from "@/modules/Cards/Card";
import { Rank } from "@/modules/Cards/Rank";
import { Suit } from "@/modules/Cards/Suit";
import { compareArrays } from "@/modules/Collections";

export enum Combination {
	Nothing,
	JacksOrBetter,
	TwoPairs,
	ThreeOfAKind,
	Straight,
	Flush,
	FullHouse,
	FourOfAKind,
	StraightFlush,
	RoyalFlush
}

export interface BestCombination {
	combination: Combination;
	cards: Card[]
}

interface GroupedCards extends Array<Array<Card>> {}

const fiveHigh = [Rank.Ace, Rank.Two, Rank.Three, Rank.Four, Rank.Five];
const aceHigh = [Rank.Ace, Rank.Ten, Rank.Jack, Rank.Queen, Rank.King];
const nothing: BestCombination = { 
	combination: Combination.Nothing, 
	cards: null 
};

export function getBestCombination(input: Card[]): BestCombination {
	// Check hand
	if (!handIsValid(input)) return nothing;
	
	// Sort input hand by rank, removing jokers
	let hand = input.filter(x => x.suit != Suit.Joker)
		.sort((a, b) => a.rank - b.rank);

	// Straight and flush results are used for multiple combinations
	let straight = isAcesStraight(hand) || isStraight(hand);
	let flush = isFlush(hand);

	// Check royal/straight flush
	let result = isRoyalFlush(straight, flush) || isStraightFlush(straight, flush);

	// More data is needed
	if (!result) {
		// Group cards by rank to find pairs etc
		let groupedCards = groupCardsByRank(hand);

		// Check remaining combinations
		result = isFourOfAKind(groupedCards)
			|| isFullHouse(groupedCards)
			|| flush
			|| straight
			|| isThreeOfAKind(groupedCards)
			|| isTwoPairs(groupedCards)
			|| isJacksOrBetter(groupedCards)
			|| nothing;
	}

	// Finished
	return result;
}

function handIsValid(hand: Card[]) {
	// Check array
	if (hand == null) return false;

	// Filter jokers, create array of card ids
	let ids = hand.filter(x => x.suit != Suit.Joker)
		.map(x => getCardIndex(x));

	// Check for duplicate cards
	let uniques = new Set(ids);

	// Should have at least 2 cards
	return uniques.size == ids.length && uniques.size >= 2;
}

// Straight = five-high or aces-high
function isAcesStraight(hand: Card[]): BestCombination {
	// Ensure hand is 5 cards
	if (hand.length < 5) return;

	// Get ranks for valid hands
	const handRanks = hand.map(x => x.rank);

	// Check hand
	if (compareArrays(handRanks, fiveHigh) || compareArrays(handRanks, aceHigh)) {
		return {
			combination: Combination.Straight,
			cards: [...hand]
		};
	}
}

// Straight = any five cards of consecutive value, any suit
function isStraight(hand: Card[]): BestCombination {
	// Ensure hand is 5 cards
	if (hand.length < 5) return;

	// Compare to previous card rank
	for (let idx = 1; idx < hand.length; idx++) {
		// Get current rank
		let prev = hand[idx - 1].rank;
		let curr = hand[idx].rank;
		
		// Exit early if current card is not 1 higher than previous
		if (curr - prev != 1) return null;
	}

	// Success
	return {
		combination: Combination.Straight,
		cards: [...hand]
	};
}

// Flush = any five cards of the same suit, any order
function isFlush(hand: Card[]): BestCombination {
	// Ensure hand is 5 cards
	if (hand.length < 5) return;

	// Get suit of first card
	let suit = hand[0].suit;

	// Compare all cards to first
	if (hand.every(x => x.suit == suit)) {
		return {
			combination: Combination.Flush,
			cards: [...hand]
		};
	}
}

// Straight Flush = five consecutive cards of the same suit
function isStraightFlush(straight: BestCombination, flush: BestCombination): BestCombination {
	if (straight && flush) {
		return {
			combination: Combination.StraightFlush,
			cards: straight.cards
		};
	}
}

// Royal Flush = Ten, a Jack, a Queen, a King, and an Ace of the same suit
function isRoyalFlush(straight: BestCombination, flush: BestCombination): BestCombination {
	if (straight && flush && straight.cards[4].rank == Rank.King) {
		return {
			combination: Combination.RoyalFlush,
			cards: straight.cards
		};
	}
}

// Four of a Kind = four cards of the same value
function isFourOfAKind(groupedCards: GroupedCards): BestCombination {
	if (groupedCards[0].length == 4) {
		return {
			combination: Combination.FourOfAKind,
			cards: [...groupedCards[0]]
		};
	}
}

// Full House = a pair and a three of a kind
function isFullHouse(groupedCards: GroupedCards): BestCombination {
	if (groupedCards.length == 2 && groupedCards[0].length == 3 && groupedCards[1].length == 2) {
		return {
			combination: Combination.FullHouse,
			cards: [...groupedCards[0], ...groupedCards[1]]
		};
	}
}

// Three of a Kind = three cards same value
function isThreeOfAKind(groupedCards: GroupedCards): BestCombination {
	if (groupedCards[0].length == 3) {
		return {
			combination: Combination.ThreeOfAKind,
			cards: [...groupedCards[0]]
		};
	}
}

// Two Pair = a pair is two cards with same value, two pairs
function isTwoPairs(groupedCards: GroupedCards): BestCombination {
	if (groupedCards.length >= 2 && groupedCards[0].length == 2 && groupedCards[1].length == 2) {
		return {
			combination: Combination.TwoPairs,
			cards: [...groupedCards[0], ...groupedCards[1]]
		};
	}
}

// Jacks or Better = a pair of high cards (J/Q/K/A)
function isJacksOrBetter(groupedCards: GroupedCards): BestCombination {
	if (groupedCards[0].length == 2) {
		let firstCardRank = groupedCards[0][0].rank;
		if (firstCardRank >= Rank.Jack || firstCardRank == Rank.Ace) {
			return {
				combination: Combination.JacksOrBetter,
				cards: [...groupedCards[0]]
			};
		}
	} 
}

// Groups cards by counting same ranks, sorted by most to least frequent
function groupCardsByRank(hand: Card[]): GroupedCards {
	let result: GroupedCards = [];
	
	for (let card of hand) {
		let existing = result.find(x => x[0].rank == card.rank);
		if (existing) existing.push(card);
		else result.push([card]);
	}
	
	return result.sort((a, b) => b.length - a.length);
}

export const scoringInternal = {
	handIsValid, 
	isAcesStraight, 
	isStraight, 
	isFlush, 
	isStraightFlush, 
	isRoyalFlush, 
	isFourOfAKind, 
	isFullHouse, 
	isThreeOfAKind, 
	isTwoPairs,
	isJacksOrBetter,
	groupCardsByRank
};