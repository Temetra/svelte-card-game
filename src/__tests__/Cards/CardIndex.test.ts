import { Ranks, Rank } from "@/modules/Cards/Rank";
import { Suits, Suit } from "@/modules/Cards/Suit";
import { getCardIndex } from "@/modules/Cards/Card";

test("Unique indices for cards", () => {
	let indices = Suits.map(suit => {
		return Ranks.map(rank => getCardIndex(rank, suit));
	});

	expect(indices).toMatchSnapshot();
});

test("getCardIndex rank suit", () => {
	let rankSuitValue = getCardIndex(Rank.Four, Suit.Diamonds);
	expect(rankSuitValue).toBe(204);
});

test("getCardIndex card", () => {
	let rankSuitValue = getCardIndex({ rank:Rank.Four, suit:Suit.Diamonds });
	expect(rankSuitValue).toBe(204);
});
