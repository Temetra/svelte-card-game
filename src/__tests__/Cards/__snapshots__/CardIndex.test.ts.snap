// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unique indices for cards 1`] = `
Array [
  Array [
    101,
    102,
    103,
    104,
    105,
    106,
    107,
    108,
    109,
    110,
    112,
    113,
    114,
  ],
  Array [
    201,
    202,
    203,
    204,
    205,
    206,
    207,
    208,
    209,
    210,
    212,
    213,
    214,
  ],
  Array [
    301,
    302,
    303,
    304,
    305,
    306,
    307,
    308,
    309,
    310,
    312,
    313,
    314,
  ],
  Array [
    401,
    402,
    403,
    404,
    405,
    406,
    407,
    408,
    409,
    410,
    412,
    413,
    414,
  ],
]
`;
